### PR TITLE
TES optimization

### PR DESCRIPTION
**Pull request by AbbyGi**

### CODE REVIEW

File: run_optimization.py [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+                                  'shape']
+
+
+def main():
+    diff_ev(bounds=[(1000, 10000), (5, 10)], fields=field_list, popsize=5,
+        crosspb=0.8, mut=0.1, threshold=0, mut_type='rand/1')
+
+
+if __name__ == '__main__':
+    main()
```
Comment (mrakitin) [inline run_optimization.py:373]: This file may need to belong to a separate, more generic library, which can be used for optimization using both the real hardware and the simulation environment.

File: sirepo_bluesky.py [hunk 2]
```diff
@@ -103,6 +103,12 @@ def find_element(elements, field, value):
                 return e
         assert False, 'element not found, {}={}'.format(field, value)
 
+    def find_optic_id_by_name(self, optic_name):
+        for i in range(len(self.data['models']['beamline'])):
```
Comment (mrakitin) [inline sirepo_bluesky.py]:
```suggestion
        for optic_id in range(len(self.data['models']['beamline'])):
```

File: sirepo_bluesky.py [hunk 3]
```diff
@@ -103,6 +103,12 @@ def find_element(elements, field, value):
...
+            if self.data['models']['beamline'][i]['title'] == optic_name:
```
Comment (mrakitin) [inline sirepo_bluesky.py]:
```suggestion
            if self.data['models']['beamline'][optic_id]['title'] == optic_name:
```

File: sirepo_bluesky.py [hunk 4]
```diff
@@ -103,6 +103,12 @@ def find_element(elements, field, value):
...
+                return i
```
Comment (mrakitin) [inline sirepo_bluesky.py]:
```suggestion
                return optic_id
```

File: sirepo_detector.py [hunk 5]
```diff
@@ -93,11 +92,13 @@ def update_parameters(self):
...
         self.data = data
         for key, value in self.sirepo_components.items():
-            optic_id = self.find_optic_id_by_name(key, self.data)
+            optic_id = self.sb.find_optic_id_by_name(key)
             self.parameters = {f'sirepo_{k}': v for k, v in
                                data['models']['beamline'][optic_id].items()}
             for k, v in self.parameters.items():
                 getattr(value, k).set(v)
+                # data['models']['beamline']['<OPTICAL ELEMENT>']['<ITS COMPONENT>']
+                print()
```
Comment (mrakitin) [inline sirepo_detector.py]: We may not need this print.
```suggestion
```

File: sirepo_detector.py [hunk 6]
```diff
@@ -133,6 +131,7 @@ def trigger(self):
         with open(srw_file, 'wb') as f:
             f.write(self.sb.get_datafile())
 
+
```
Comment (mrakitin) [inline sirepo_detector.py]:
```suggestion
```

File: sirepo_detector.py [hunk 7]
```diff
@@ -257,12 +249,18 @@ def select_optic(self, name):
...
         real_name = "sirepo_" + name
-        if self.field0 is None:
-            self.field0 = real_name
-        else:
-            self.field1 = real_name
+        ct = 0
+        while 'field' + str(ct) in self.fields.keys():
+            ct += 1
+        fieldkey = 'field' + str(ct)
+        parentkey = 'par' + str(ct)
```
Comment (mrakitin) [inline sirepo_detector.py]: f-strings may be better in the 4 lines above. That way you will not need to wrap `ct` into `str(...)`

File: sirepo_detector.py [hunk 8]
```diff
@@ -257,12 +249,18 @@ def select_optic(self, name):
...
+
+        self.fields[fieldkey] = real_name
+        self.parents[parentkey] = self.sirepo_component.name
+        key = self.parents[parentkey] + '_' + name
```
Comment (mrakitin) [inline sirepo_detector.py]: f-string here too.

File: sirepo_flyer.py [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+from collections import deque
+from multiprocessing import Process
+from pathlib import Path
+
+from ophyd.sim import NullStatus, new_uid
+
+from sirepo_bluesky import SirepoBluesky
+from srw_handler import read_srw_file
+
+from re_config import *
```
Comment (mrakitin) [inline sirepo_flyer.py]: Let's move this line in the `if __name__ == '__main__'` block, and comment it out.

File: sirepo_flyer.py [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+    def __init__(self, sim_id, server_name, params_to_change, sim_code='srw', copy_count=5, watch_name='Watchpoint', run_parallel=True):
+        super().__init__()
+        self.name = 'sirepo_flyer'
+        self.sim_id = sim_id
+        self.server_name = server_name
+        self.params_to_change = params_to_change
+        self.sim_code = sim_code
+        self._copy_count = copy_count
+        self.watch_name = watch_name
+        self.run_parallel = run_parallel
```
Comment (mrakitin) [inline sirepo_flyer.py]: Most of the attributes may need to be properties too.

File: sirepo_flyer.py [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+                        'root': '/tmp/data',  # from 00-startup.py (added by mrakitin for future generations :D)
+                        'resource_path': srw_file,
+                        'resource_kwargs': {},
+                        'path_semantics': {'posix': 'posix', 'nt': 'windows'}[os.name],
+                        'uid': _resource_uid}
+            self._resource_uids.append(_resource_uid)
+            self._asset_docs_cache.append(('resource', resource))
+
+        for param in self.params_to_change:
+            # for i in range(self.copy_count):
```
Comment (mrakitin) [inline sirepo_flyer.py]:
```suggestion
```

File: sirepo_flyer.py [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+            # name doesn't need to be unique, server will rename it
+            c1 = sb.copy_sim('{} Bluesky'.format(sb.data['models']['simulation']['name']), )
+            print('copy {}, {}'.format(c1.sim_id, c1.data['models']['simulation']['name']))
+            # want to be able to change any parameters to any value
```
Comment (mrakitin) [inline sirepo_flyer.py]:
```suggestion
```

File: sirepo_flyer.py [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+                p = Process(target=self._run, args=(self._copies[i],))
+                p.start()
+                procs.append(p)
+            # wait for procs to finish
+            for p in procs:
+                p.join()
+        else:
+            # run serial
+            for i in range(self.copy_count):
+                print('running sim: {}', self._copies[i].sim_id)
```
Comment (mrakitin) [inline sirepo_flyer.py]:
```suggestion
                print(f'running sim: {self._copies[i].sim_id}')
```

File: sirepo_flyer.py [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+                status = self._copies[i].run_simulation()
+                print('Status:', status['state'])
```
Comment (mrakitin) [inline sirepo_flyer.py:182]: We need to have this status stored in databroker.

File: sirepo_flyer.py [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+                c1.data['models']['beamline'][optic_id].update(parameters_to_update)
+            watch = sb.find_element(c1.data['models']['beamline'], 'title', self.watch_name)
+            c1.data['report'] = 'watchpointReport{}'.format(watch['id'])
+            self._copies.append(c1)
+            print('COPIES', len(self._copies))
+
+        if self.run_parallel:
+            procs = []
+            for i in range(self.copy_count):
+                p = Process(target=self._run, args=(self._copies[i],))
```
Comment (mrakitin) [inline sirepo_flyer.py]: Need to figure out how we can pass the status from `self._run`.

File: sirepo_flyer.py [hunk 16]
```diff
@@ -0,0 +1,241 @@
...
+                     'datum_id': datum_id}
+            self._asset_docs_cache.append(('datum', datum))
+            self._datum_ids.append(datum_id)
+        return NullStatus()
+
+    def describe_collect(self):
+        elem_name = []
+        curr_param = []
+        for i in range(len(self.params_to_change)):
+            inputs = self.params_to_change[i]
```
Comment (mrakitin) [inline sirepo_flyer.py]: I think we can just loop over `self.params_to_change` and get `inputs` in the for loop.

File: sirepo_flyer.py [hunk 17]
```diff
@@ -0,0 +1,241 @@
...
+                curr_param.append(list(parameters_to_update.keys()))  # e.g., 'horizontalSize'
+
+        return_dict = {self.name:
+                        {f'{self.name}_image': {'source': f'{self.name}_image',
+                                                'dtype': 'array',
+                                                'shape': [-1, -1],
+                                                'external': 'FILESTORE:'},
+                         f'{self.name}_shape': {'source': f'{self.name}_shape',
+                                               'dtype': 'array',
+                                               'shape': [1]},
```
Comment (mrakitin) [inline sirepo_flyer.py]: Should not it be `2`?

File: sirepo_flyer.py [hunk 18]
```diff
@@ -0,0 +1,241 @@
...
+                                               'dtype': 'array',
+                                               'shape': [2]},
+                         f'{self.name}_vertical_extent': {'source': f'{self.name}_vertical_extent',
+                                               'dtype': 'array',
+                                               'shape': [2]},
+                         }
+                       }
+        for i in range(len(elem_name)):
+            for j in range(len(curr_param[i])):
+                return_dict[self.name][f'{self.name}_{elem_name[i]}_{curr_param[i][j]}'] = {'source': f'{self.name}_{elem_name[i]}_{curr_param[i][j]}',
```
Comment (mrakitin) [inline sirepo_flyer.py]: The line is too long, can be split into two lines for better readability.

File: sirepo_flyer.py [hunk 19]
```diff
@@ -0,0 +1,241 @@
...
+                f.write(data_file)
+
+            ret = read_srw_file(self._srw_files[i])
+            means.append(ret['mean'])
+            shapes.append(ret['shape'])
+            photon_energies.append(ret['photon_energy'])
+            horizontal_extents.append(ret['horizontal_extent'])
+            vertical_extents.append(ret['vertical_extent'])
+
+            print('copy {} data hash: {}'.format(self._copies[i].sim_id, hashlib.md5(data_file).hexdigest()))
```
Comment (mrakitin) [inline sirepo_flyer.py]: Let's keep these hash values and save them in databroker to validate the files later in the future.

File: sirepo_flyer.py [hunk 20]
```diff
@@ -0,0 +1,241 @@
...
+            self._copies[i].delete_copy()
+
+        assert len(self._copies) == len(self._datum_ids), \
+            f'len(self._copies) != len(self._datum_ids) ({len(self._copies)} != {len(self._datum_ids)})'
+
+        now = ttime.time()
+        for i, datum_id in enumerate(self._datum_ids):
+            elem_name = []
+            curr_param = []
```
Comment (mrakitin) [inline sirepo_flyer.py:275]: These 2 lists can be better moved next to the for loops below.

File: sirepo_flyer.py [hunk 21]
```diff
@@ -0,0 +1,241 @@
...
+                inputs = self.params_to_change[j]
+                for key, parameters_to_update in inputs.items():
+                    elem_name.append(key)  # e.g., 'Aperture'
+                    curr_param.append(list(parameters_to_update.keys()))  # e.g., 'horizontalSize'
+
+            for ii in range(len(elem_name)):
+                for jj in range(len(curr_param[ii])):
+                    data[f'{self.name}_{elem_name[ii]}_{curr_param[ii][jj]}'] = params_to_change[i][elem_name[ii]][curr_param[ii][jj]]
+                    # params_to_change[2][elem_name[ii]][curr_param[ii][jj]]
+            yield {'data': data,
```
Comment (mrakitin) [inline sirepo_flyer.py:297]:
```suggestion

            yield {'data': data,
```

File: sirepo_flyer.py [hunk 22]
```diff
@@ -0,0 +1,241 @@
...
+            for j in range(len(self.params_to_change)):
+                inputs = self.params_to_change[j]
+                for key, parameters_to_update in inputs.items():
+                    elem_name.append(key)  # e.g., 'Aperture'
+                    curr_param.append(list(parameters_to_update.keys()))  # e.g., 'horizontalSize'
+
+            for ii in range(len(elem_name)):
+                for jj in range(len(curr_param[ii])):
+                    data[f'{self.name}_{elem_name[ii]}_{curr_param[ii][jj]}'] = params_to_change[i][elem_name[ii]][curr_param[ii][jj]]
+                    # params_to_change[2][elem_name[ii]][curr_param[ii][jj]]
```
Comment (mrakitin) [inline sirepo_flyer.py]:
```suggestion
```

File: sirepo_flyer.py [hunk 23]
```diff
@@ -0,0 +1,241 @@
...
+            yield {'data': data,
+                   'timestamps': {key: now for key in data}, 'time': now,
+                   'filled': {key: False for key in data}}
+
+    def _run(self, sim):
+        print('running sim {}'.format(sim.sim_id))
+        status = sim.run_simulation()
+        print('Status:', status['state'])
```
Comment (mrakitin) [inline sirepo_flyer.py:305]:
```suggestion
        print('Status:', status['state'])
        return status
```

File: sirepo_flyer.py [hunk 24]
```diff
@@ -0,0 +1,241 @@
...
+        params_to_change.append({key1: parameters_update1,
+                                 key2: parameters_update2})
+    print('Length of params_to_change', len(params_to_change), 'outside class')
+
+    sirepo_flyer = SirepoFlyer(sim_id='87XJ4oEb', server_name='http://10.10.10.10:8000', params_to_change=params_to_change,
+                               watch_name='W60')
+
+    RE(bp.fly([sirepo_flyer]))
+
+    # sirepo_flyer = SirepoFlyer(sim_id='87XJ4oEb', server_name='http://10.10.10.10:8000')
```
Comment (mrakitin) [inline sirepo_flyer.py]:
```suggestion
```

File: sirepo_flyer.py [hunk 25]
```diff
@@ -0,0 +1,241 @@
...
+            elem_name = []
+            curr_param = []
+
+            data = {f'{self.name}_image': datum_id,
+                    f'{self.name}_shape': shapes[i],
+                    f'{self.name}_mean': means[i],
+                    f'{self.name}_photon_energy': photon_energies[i],
+                    f'{self.name}_horizontal_extent': horizontal_extents[i],
+                    f'{self.name}_vertical_extent': vertical_extents[i],
+                    }
```
Comment (mrakitin) [inline sirepo_flyer.py:284]: We also need to save the hash and the status of the simulation.

**Review by mrakitin — COMMENTED** [review]
Thanks @AbbyGi! I have a few small suggestions (see below).

**Review by mrakitin — APPROVED** [review]
This is in a great shape, even with tests! Ready to go to the cookiecutter-based repo (to be created.)

**Comment by mrakitin** [comment]
@AbbyGi, do we still need it? Or were all needed parts moved to https://github.com/NSLS-II/sirepo-bluesky (a proper package)?

**Comment by AbbyGi** [comment]
All necessary code was moved, so we shouldn't need this anymore.

**Comment by mrakitin** [comment]
OK, sounds good. Closing.

**Pull request closed without merging**